### ISSUE TITLE
"Current thread does not hold the state lock for project" in a simple multi-project build

### ISSUE THREAD

**Issue by netvl** [gradle/gradle#15730]
When `org.gradle.parallel=true` is set, a simple multi-project Gradle build using the Scoverage plugin works in Gradle 6.6.1 and does not work in Gradle 6.7/6.7.1.

Disabling parallel build makes the build work on 6.7+ as well.

Please note that while this does affect those builds which use the Scoverage plugin, I'm submitting this ticket here and not in the Scoverage tracker because the problem appears after the Gradle upgrade, so it seems like a Gradle regression. Please do tell me if I'm wrong.

### Expected Behavior

A simple multi-project build using the Scoverage plugin works.

### Current Behavior

When parallel build is enabled, the build works correctly under Gradle 6.6.1 and fails under Gradle 6.7/6.7.1 with the following error:

```
A problem occurred configuring project ':first'.
> Could not determine the dependencies of task ':second:compileJava'.
   > Current thread does not hold the state lock for project :second
```

Disabling `org.gradle.parallel` makes the build work again under Gradle 6.7+, but for a large project parallel building really helps with performance, so this is not a good workaround.

### Context

This worked on Gradle 6.6.1 and stopped working in Gradle 6.7+. It affected my complex multi-project build which relies on the Scoverage plugin, but the only thing I have changed is that I bumped the Gradle version and nothing else.

### Steps to Reproduce 

Please find an example project attached: [example.zip](https://github.com/gradle/gradle/files/5790170/example.zip)

### Your Environment

Build scan URL:

**Comment by netvl** [comment]
Could be related to https://github.com/gradle/gradle/issues/14516

**Comment by vlsi** [comment]
```
Caused by: org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':second:compileJava'.
        at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.getDependencies(CachingTaskDependencyResolveContext.java:69)
        at org.gradle.api.internal.tasks.AbstractTaskDependency.getDependencies(AbstractTaskDependency.java:47)
        at org.gradle.api.tasks.TaskDependency$getDependencies.call(Unknown Source)
        at org.gradle.api.tasks.TaskDependency$getDependencies.call(Unknown Source)
        at org.scoverage.ScoveragePlugin.recursiveDependenciesOf(ScoveragePlugin.groovy:376)
```

This looks like `gradle-scoverage` issue to me.

scoverage performs https://github.com/scoverage/gradle-scoverage/blob/f92c9424ad8b02e13dcbfc790c8995c51fd1b308/src/main/groovy/org/scoverage/ScoveragePlugin.groovy#L156-L159

```
project.afterEvaluate {
    def compileTask = project.tasks[instrumentedSourceSet.getCompileTaskName("scala")]
    // later in recursiveDependenciesOf:
    task.getTaskDependencies().getDependencies(task) // <-- fails here
}
```

It looks like cross-project configuration which has always been discouraged in Gradle.

Have you asked Scoverage team for the analysis?

**Comment by netvl** [comment]
I haven't (I probably will, to speed up the resolution), but at this particular point there is no cross-project access, as far as I can see. Even in your snippet, `project` is the project the plugin is being applied to, and `compileTask` is the compilation task within this project, and then `recursiveDependenciesOf` gets dependencies of that task. I don't see where cross-project access happens.

**Comment by netvl** [comment]
Anyway, gradle-scoverage does not appear to use any non-public API, and the only thing that changed and resulted in this error is the Gradle version bump, and not even in the major version. I would presume that this is very much a Gradle issue, not the plugin's.

**Comment by vlsi** [comment]
> gradle-scoverage does not appear to use any non-public API

Could you please reproduce the issue without gradle-scoverage plugin by using Gradle APIs only?

If you require `gradle-scoverage` plugin for the issue to appear, then please analyze `gradle-scoverage` or ask someone from `gradle-scoverage` team to help you.

> but at this particular point there is no cross-project access

Cross-project (or something of that kind) access happens when `recursiveDependenciesOf` enters recursion

**Comment by netvl** [comment]
Sure, I'm totally able to reproduce this in custom code, without any plugins. Please find example [attached](https://github.com/gradle/gradle/files/5800381/example.zip). Just for an easier reference, here is the code of the example plugin which I wrote instead of Scoverage, it is even simpler since it does not add any tasks - the error can be replicated with built-in tasks just fine:

```kotlin
package build

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.tasks.scala.ScalaCompile
import org.gradle.kotlin.dsl.*

open class ExamplePlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.apply(plugin = "scala")

        val compileTask = project.tasks.getByName("compileScala")

        project.afterEvaluate {
            val dependencies = recursiveDependenciesOf(compileTask).filter { it is ScalaCompile }
        }
    }

    private val taskDependencies: MutableMap<Task, Set<Task>> = mutableMapOf()

    private fun recursiveDependenciesOf(task: Task): Set<Task> {
        if (!taskDependencies.containsKey(task)) {
            val directDependencies = task.taskDependencies.getDependencies(task).toSet()
            val nestedDependencies = directDependencies.flatMap { recursiveDependenciesOf(it) }.toSet()
            val dependencies = directDependencies + nestedDependencies
            taskDependencies[task] = dependencies
            return dependencies
        } else {
            return taskDependencies[task]!!
        }
    }
}
```

> Cross-project (or something of that kind) access happens when recursiveDependenciesOf enters recursion

I'm sorry but I fail to see how this could happen. In this example, the `:second` project does not have any dependencies on the `:first` project, so there is no cross-project access, everything is limited to just the `:second` project. In the example above, **no** non-public API is used, and it is totally unfair to say that analyzing a set of dependencies of a task is necessarily cross-project or something of that kind - I imagine there could be a lot of reasons a plugin would want to analyze a task graph.

**Comment by vlsi** [comment]
Thanks for the smaller example.

> so there is no cross-project access, everything is limited to just the :second project

The root `build.gradle.kts` has the following, so it applies `ExamplePlugin` to **both** `first` and `second`

```kotlin
subprojects {
    apply {
        plugin("scala")
        plugin(build.ExamplePlugin::class)
    }
```

Then, `:first` project has classpath dependency on `implementation(project(":second"))`. This causes `:first:compileJava` depend on `:second:compileJava`, which later triggers `Current thread does not hold the state`.

If you remove `implementation(project(":second"))` from `first/build.gradle.kts`, then the error disappears.

Note: even though configuration phase is single-threaded only in the current Gradle versions, the order of sibling project evaluation is not defined, so configuration code should not span across project boundaries.

**Comment by netvl** [comment]
Oh, I see it now - I was confused by the fact that it printed "Could not determine the dependencies of task ':**second:**compileJava'", so I thought that the plugin got applied to the second project only. But yes, since it the `:second` is a dependency of `:first`, it is actually thrown from the `:first` project - doing something like
```
    apply {
        plugin("scala")
        if (name == "second") {
            plugin(build.ExamplePlugin::class)
        }
    }
```
also solves the problem.

Thanks for the explanation!

However, what would be the solution here? I believe doing something like the Scoverage plugin does still makes sense in some cases, and in general, computing a list of task dependencies can be useful for build analysis plugins or something like that.

**Comment by vlsi** [comment]
> However, what would be the solution here?

I do not know what scoverage is doing, so I can't suggest the solution.

> can be useful for build analysis plugins or something like that

It is hard to suggest something without a use case. Please share the use case.

> computing a list of task dependencies can be useful for build analysis plugins or something like that

I believe the common case is documented at https://docs.gradle.org/current/userguide/cross_project_publications.html
If something should be shared, then it should be shared with configurations.

In general, it is invalid to assume that "task dependency information is available". What if a new task is added after the analysis?
In other words, suppose scoverage computed the dependencies and configured its tasks accordingly. What if a new dependency or task is added later?

On top of that, projects might evaluate in a different order, so inspecting "other" projects is not safe either.
Note: there's no "proper" order, so it is better to assume the order does not exist, and it is better to avoid cross-project access.

**Comment by vlsi** [comment]
Should the issue be closed then?

**Comment by dsilvasc** [comment]
If it's unsafe to access this public api (is the problem `task.taskDependencies` or `task.taskDependencies.getDependencies(task)`?), then should it be removed?  Should it throw an exception explaining the issue when called in a cross-project way?

Right now it looks like gradle crashes with an internal exception.

**Comment by eyalroth** [comment]
Hey, I'm one of the contributors of `gradle-scoverage` (sorry for the late response). The reason that the plugin recursively discovers task dependencies deeper than one level is [described here](https://github.com/scoverage/gradle-scoverage/issues/150#issuecomment-903091145).

For now, I'm planning to change the plugin so it will perform this discovery only if the "no normal compilation" is enabled. This means that only this mode will be incompatible with parallel builds, which is still a bit unfortunate, but a fair compromise given that the mode purpose to improve build times (which is also the purpose of parallel builds).

> It looks like cross-project configuration which has always been discouraged in Gradle.

> Note: even though configuration phase is single-threaded only in the current Gradle versions, the order of sibling project evaluation is not defined, so configuration code should not span across project boundaries.

This is unfortunate, as we would love to use this functionality in the scoverage plugin.

> I do not know what scoverage is doing, so I can't suggest the solution.

By default, the scoverage plugin creates an additional source set named `scoverage` with its own (scala) compilation task with the scoverage compiler plugin, which in turn instruments the compiled classes with coverage code.

Running any coverage task means that both the normal compilation task (`compileScoverage`) and the additional compilation task (`compileScoverageScala`) will run. When a process intends to both report/check coverage and publish the code, this makes sense. However, when one wishes to only report/check coverage and is not concerned about the final compiled classes having instrumented code in them, the double compilation is redundant and costs time/resources.

Therefore, the "run without normal compilation" mode/feature was introduced into the plugin. In this mode, the plugin needs to make sure that the scoverage compiled classes are available to any other task that relies on the compiled output of the project; i.e, the compilation tasks of any other project that depends on the current project.

In order to achieve this, the plugin points the scoverage compilation task destination dir to the destination dir of the normal compilation, and also make every scoverage compilation task depend on the scoverage compilation tasks of the projects that the normal compilation task depends on (this is where the recursive dependency discovery comes into place).

For example, in the default mode we have this dependency graph (assuming `common` is a dependency of `a`):
```
:common:reportScoverage --> :common:compileScoverageScala --> :common:compileScala
:a:reportScoverage --> :a:compileScoverageScala --> :a:compileScala --^ 
```
While in the "without normal compilation" mode we have this:
```
:common:reportScoverage --> :common:compileScoverageScala --> :common:compileScala
:a:reportScoverage --> :a:compileScoverageScala --^--> :a:compileScala --^ 
```
(`:a:compileScoverageScala` depends on both `:common:compileScoverageScala` and `:a:compileScala`).

Furthermore, in this mode, since the output jars of projects consist of instrumented classes instead of normally compiled classes (like in the default mode), the plugin needs to make sure that tests -- which use the output jars of their dependent projects -- run after the report task of their dependent projects, so that they will not affect the coverage of their dependencies.

**Comment by eyalroth** [comment]
> I know there are things we could do better here to make it easier for plugins to swap out outputs or insert themselves "between" two tasks, but we're not going to soften/remove the error that's trying to prevent one project from reaching into others and so I'm going to close this issue for now.

@big-guy That's completely understandable, and as I mentioned earlier -- it's not all that devastating for the plugin since it only affects the "no normal compilation" mode.

> IIUC, you want to get the instrumented code from subprojects so that when you aggregate the reports, you count all possible coverage regardless of which subproject had the tests.

It's actually the other way around. The coverage of each subproject is isolated from the tests of other projects. That's why in the "no normal compilation" mode, the plugin needs to make sure to run the coverage report task of a subproject before running the tests of all the other projects that depend on it (either directly or via proxy dependency). 

> 1. Don't create the separate instrumented compilation tasks and just directly configure the "normal" Scala compile tasks to produce instrumented code when instrumentation is needed.

This won't help making sure that the report task runs before other dependent projects' tests.

> 2. Create a parallel set of tasks for compilation that do instrumentation but do not interact with the "normal" tasks. e.g., you would not write the instrumented Scala code into the same location as the original Scala compile task. It should be possible to then swap between instrumented/non-instrumented, but it's slightly complicated to hook up because you'll need to also decide if you're going to run tests with instrumented code all the time or not.

That's what the plugin currently does in its default mode, but again, it doesn't help with the aforementioned problem
